ProcessLib/BoundaryCondition/Python/PythonBoundaryCondition.cpp

```diff
@@ -0,0 +1,239 @@
+    }
+
+    //! Optionally flushes Python's stdout after running Python code.
+    ~FlushStdoutGuard()
+    {
+        if (!_flush)
+            return;
+
+        using namespace pybind11::literals;
+        pybind11::print("end"_a = "", "flush"_a = true);
+    }
+
+private:
+    //! To flush or not to flush.
+    const bool _flush;
+};
+}  // anonymous namespace
+
+namespace ProcessLib
+{
+PythonBoundaryCondition::PythonBoundaryCondition(
+    PythonBoundaryConditionData&& bc_data,
+    unsigned const integration_order,
+    unsigned const shapefunction_order,
+    unsigned const global_dim,
+    bool const flush_stdout)
+    : _bc_data(std::move(bc_data)), _flush_stdout(flush_stdout)
+{
+    std::vector<MeshLib::Node*> const& bc_nodes =
+        _bc_data.boundary_mesh.getNodes();
+    MeshLib::MeshSubset bc_mesh_subset(_bc_data.boundary_mesh, bc_nodes);
+
+    // Create local DOF table from the bc mesh subset for the given variable and
+    // component id.
+    _dof_table_boundary = _bc_data.dof_table_bulk.deriveBoundaryConstrainedMap(
+        std::move(bc_mesh_subset));
+
+    createLocalAssemblers<PythonBoundaryConditionLocalAssembler>(
+        global_dim, _bc_data.boundary_mesh.getElements(), *_dof_table_boundary,
+        shapefunction_order, _local_assemblers,
+        _bc_data.boundary_mesh.isAxiallySymmetric(), integration_order,
+        _bc_data);
+}
+
+void PythonBoundaryCondition::getEssentialBCValues(
+    const double t, GlobalVector const& x,
+    NumLib::IndexValueVector<GlobalIndexType>& bc_values) const
+{
+    FlushStdoutGuard guard(_flush_stdout);
+    (void)guard;
+
+    auto const nodes = _bc_data.boundary_mesh.getNodes();
+
+    auto const& bulk_node_ids_map =
+        *_bc_data.boundary_mesh.getProperties().getPropertyVector<std::size_t>(
+            "bulk_node_ids");
+
+    bc_values.ids.clear();
+    bc_values.values.clear();
+
+    bc_values.ids.reserve(_bc_data.boundary_mesh.getNumberOfNodes());
+    bc_values.values.reserve(_bc_data.boundary_mesh.getNumberOfNodes());
+
+    std::vector<double> primary_variables;
+
+    for (auto const* node : _bc_data.boundary_mesh.getNodes())
+    {
+        auto const boundary_node_id = node->getID();
+        auto const bulk_node_id = bulk_node_ids_map[boundary_node_id];
+
+        // gather primary variable values
+        primary_variables.clear();
+        auto const num_var = _dof_table_boundary->getNumberOfVariables();
+        for (int var = 0; var < num_var; ++var)
+        {
+            auto const num_comp =
+                _dof_table_boundary->getNumberOfVariableComponents(var);
+            for (int comp = 0; comp < num_comp; ++comp)
+            {
+                MeshLib::Location loc{_bc_data.bulk_mesh_id,
+                                      MeshLib::MeshItemType::Node,
+                                      bulk_node_id};
+                auto const dof_idx =
+                    _bc_data.dof_table_bulk.getGlobalIndex(loc, var, comp);
```

Comment:
Is it possible to compute the `dof_idx` also from the boundary dof table?

Reply:
I think, it's not. I think I tried, but the dof table for the boundary is for a hypothetical equation system that is assembled solely on the boundary (i.e., with no connection to the bulk). Therefore all boundary dof indices would be wrong. @endJunction, is that right? If unsure, I can also check that.